### DISCUSSION TITLE
Should we account for webkitCompassHeading & webkitCompassAccuracy within the useDeviceOrientation() sensor?

### DISCUSSION THREAD

**Discussion by michealroberts** [vueuse/vueuse#787] [Ideas]
I've been thinking. There are two unofficial properties on the DeviceOrientationEvent that are available in iOS Safari/ webkit based browsers.

From [Mozilla Web Docs DeviceOrientationEvent](https://developer.mozilla.org/en-US/docs/Web/API/DeviceOrientationEvent):

**DeviceOrientationEvent.webkitCompassHeading**  [Read only]

A number represents the difference between the motion of the device around the z axis of the world system and the direction of the north, express in degrees with values ranging from 0 to 360.

**DeviceOrientationEvent.webkitCompassAccuracy**  [Read only]
The accuracy of the compass means that the deviation is positive or negative. It's usually 10.

---

So they're not in the official specification for the device orientation web API, but they're available nonetheless (and extremely useful for calculating the actual alpha compass heading value on iOS devices).

I would like to propose modification of packages/core/useDeviceOrientation/index.ts to the following:

```
/* this implementation is original ported from https://github.com/logaretm/vue-use-web by Abdelrahman Awad */

import { ref, Ref } from 'vue-demi'
import { useEventListener } from '../useEventListener'
import { ConfigurableWindow, defaultWindow } from '../_configurable'

/**
 * Reactive DeviceOrientationEvent.
 *
 * @see https://vueuse.org/useDeviceOrientation
 * @param options
 */
export function useDeviceOrientation(options: ConfigurableWindow = {}) {
  const { window = defaultWindow } = options

  const isSupported = Boolean(window && 'DeviceOrientationEvent' in window)

  const isAbsolute = ref(false)
  const alpha: Ref<number | null> = ref(null)
  const beta: Ref<number | null> = ref(null)
  const gamma: Ref<number | null> = ref(null)
  const webkitCompassHeading: Ref<number | null> = ref(null)
  const webkitCompassAccuracy: Ref<number | null> = ref(null)

  if (window && isSupported) {
    useEventListener(window, 'deviceorientation', (event) => {
      isAbsolute.value = event.absolute
      alpha.value = event.alpha
      beta.value = event.beta
      gamma.value = event.gamma

      if (event.webkitCompassHeading) {
        webkitCompassHeading.value = event.webkitCompassHeading
      }

      if (event.webkitCompassAccuracy) {
        webkitCompassAccuracy.value = event.webkitCompassAccuracy
      }
    })
  }

  return {
    isSupported,
    isAbsolute,
    alpha,
    beta,
    gamma,
    webkitCompassHeading,
    webkitCompassAccuracy
  }
}

export type UseDeviceOrientationReturn = ReturnType<typeof useDeviceOrientation>
```

We would also need to add an extra declaration, something such as this to packages/core/useDeviceOrientation/deviceOrientationEvenet.d.ts:

```
declare interface DeviceOrientationEvent {
  webkitCompassHeading: number | null
  webkitCompassAccuracy: number | null
  requestPermission: () => string
}
```

I believe we can then utilise these two properties for iOS Safari based calculations. 

The need for this is that Android devices give an absolute alpha value, whereas iOS Safari devices give an alpha value that is relative to the device's locations when the API was first instructed.

I have the code ready to go ... I'm happy to PR this if people feel that accounting for the webkit specification is acceptable use and that they're happy to modify the global `DeviceOrientationEvent type` ...?

**Comment by michealroberts** [comment]
PR here: https://github.com/vueuse/vueuse/pull/788

**Comment by antfu** [comment]
Thanks for bringing this up. As commented in https://github.com/vueuse/vueuse/pull/788#issuecomment-925668323, I think it's too early to have this built-in. When it comes without the `webkit` prefix, I am happy to have them.

**Comment by michealroberts** [comment]
No worries @antfu I felt also that it was slightly external to the Web API standard and may lead to trouble, I think it's more of a case of Safari becoming the new IE of browsers sometimes. I've implemented my own composable version for a local project which is sufficient, I just like what you're doing here with this library. Thank you nonetheless for the quick review but also for this library as a whole!

**Comment by michealroberts** [comment]
@antfu A colleague has just made a suggestion, could we add this to some new vueuse repository that's not core but allows for non-standard implementations of the core Web APIs? Either as a package of vueuse (but not in core), or perhaps we could add a new repo a lá motion / sounds / gesture ?

**Reply by michealroberts** [reply]
Perhaps something like a vueuse/irregular sub-package?

**Reply by antfu** [reply]
Interesting thoughts! Do you have any more of these functions in mind? If we have like ten-ish functions I am happy to have a new package for experimental functions, like `@vueuse/unstable` or something that is addressed to be changed frequently as a battleground for testing experimental things.

**Reply by michealroberts** [reply]
@antfu Hmmm, well we have the GestureEvent that is webkit specific and non-standard, but then you have the gestures library ... then there's convertPointFromPageToNode which is not only non-standard but it's not on any standards track. I feel like I am now talking myself out of this concept! Haha :D
